File: packages/pose-detection/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/posedetection/classes/results/ProcessImageResult.java
package io.capawesome.capacitorjs.plugins.mlkit.posedetection.classes.results;

import androidx.annotation.NonNull;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.google.mlkit.vision.common.PointF3D;
import com.google.mlkit.vision.pose.Pose;
import com.google.mlkit.vision.pose.PoseLandmark;
import io.capawesome.capacitorjs.plugins.mlkit.posedetection.interfaces.Result;
import java.util.List;

public class ProcessImageResult implements Result {

    @NonNull
    private final Pose pose;

    public ProcessImageResult(@NonNull Pose pose) {
        this.pose = pose;
    }

    @Override
    @NonNull
    public JSObject toJSObject() {
        JSObject result = new JSObject();
        result.put("poses", this.createPosesResult(pose));
        return result;
    }

    private JSArray createPosesResult(@NonNull Pose pose) {
        JSArray result = new JSArray();
        List<PoseLandmark> landmarks = pose.getAllPoseLandmarks();
        if (!landmarks.isEmpty()) {
            JSObject poseResult = new JSObject();
            poseResult.put("landmarks", this.createLandmarksResult(landmarks));
            result.put(poseResult);
        }
        return result;
    }

    private JSArray createLandmarksResult(@NonNull List<PoseLandmark> landmarks) {
        JSArray result = new JSArray();
        for (PoseLandmark landmark : landmarks) {
            result.put(this.createLandmarkResult(landmark));
        }
        return result;
    }

    private JSObject createLandmarkResult(@NonNull PoseLandmark landmark) {
        PointF3D position = landmark.getPosition3D();
        JSObject result = new JSObject();
        result.put("type", this.convertLandmarkTypeToString(landmark.getLandmarkType()));
        result.put("x", position.getX());
        result.put("y", position.getY());
        result.put("z", position.getZ());
        result.put("inFrameLikelihood", landmark.getInFrameLikelihood());
        return result;
    }

    @NonNull
    private String convertLandmarkTypeToString(int type) {
        switch (type) {
            case PoseLandmark.NOSE:
                return "NOSE";
            case PoseLandmark.LEFT_EYE_INNER:
                return "LEFT_EYE_INNER";
            case PoseLandmark.LEFT_EYE:
                return "LEFT_EYE";
            case PoseLandmark.LEFT_EYE_OUTER:
                return "LEFT_EYE_OUTER";
            case PoseLandmark.RIGHT_EYE_INNER:
                return "RIGHT_EYE_INNER";
            case PoseLandmark.RIGHT_EYE:
                return "RIGHT_EYE";
            case PoseLandmark.RIGHT_EYE_OUTER:
                return "RIGHT_EYE_OUTER";
            case PoseLandmark.LEFT_EAR:
                return "LEFT_EAR";
            case PoseLandmark.RIGHT_EAR:
                return "RIGHT_EAR";
            case PoseLandmark.LEFT_MOUTH:
                return "LEFT_MOUTH";
            case PoseLandmark.RIGHT_MOUTH:
                return "RIGHT_MOUTH";
            case PoseLandmark.LEFT_SHOULDER:
                return "LEFT_SHOULDER";
            case PoseLandmark.RIGHT_SHOULDER:
                return "RIGHT_SHOULDER";
            case PoseLandmark.LEFT_ELBOW:
                return "LEFT_ELBOW";
            case PoseLandmark.RIGHT_ELBOW:
                return "RIGHT_ELBOW";
            case PoseLandmark.LEFT_WRIST:
                return "LEFT_WRIST";
            case PoseLandmark.RIGHT_WRIST:
                return "RIGHT_WRIST";
            case PoseLandmark.LEFT_PINKY:
                return "LEFT_PINKY";
            case PoseLandmark.RIGHT_PINKY:
                return "RIGHT_PINKY";
            case PoseLandmark.LEFT_INDEX:
                return "LEFT_INDEX";
            case PoseLandmark.RIGHT_INDEX:
                return "RIGHT_INDEX";
            case PoseLandmark.LEFT_THUMB:
                return "LEFT_THUMB";
            case PoseLandmark.RIGHT_THUMB:
                return "RIGHT_THUMB";
            case PoseLandmark.LEFT_HIP:
                return "LEFT_HIP";
            case PoseLandmark.RIGHT_HIP:
                return "RIGHT_HIP";
            case PoseLandmark.LEFT_KNEE:
                return "LEFT_KNEE";
            case PoseLandmark.RIGHT_KNEE:
                return "RIGHT_KNEE";
            case PoseLandmark.LEFT_ANKLE:
                return "LEFT_ANKLE";
            case PoseLandmark.RIGHT_ANKLE:
                return "RIGHT_ANKLE";
            case PoseLandmark.LEFT_HEEL:
                return "LEFT_HEEL";
            case PoseLandmark.RIGHT_HEEL:
                return "RIGHT_HEEL";
            case PoseLandmark.LEFT_FOOT_INDEX:
                return "LEFT_FOOT_INDEX";
            case PoseLandmark.RIGHT_FOOT_INDEX:
                return "RIGHT_FOOT_INDEX";
            default:
                return "";
        }
    }
}

File: packages/pose-detection/ios/Plugin/Classes/Results/ProcessImageResult.swift
import Foundation
import Capacitor
import MLKitPoseDetectionCommon

@objc public class ProcessImageResult: NSObject, Result {
    private static let landmarkTypeMap: [PoseLandmarkType: String] = [
        .nose: "NOSE",
        .leftEyeInner: "LEFT_EYE_INNER",
        .leftEye: "LEFT_EYE",
        .leftEyeOuter: "LEFT_EYE_OUTER",
        .rightEyeInner: "RIGHT_EYE_INNER",
        .rightEye: "RIGHT_EYE",
        .rightEyeOuter: "RIGHT_EYE_OUTER",
        .leftEar: "LEFT_EAR",
        .rightEar: "RIGHT_EAR",
        .mouthLeft: "LEFT_MOUTH",
        .mouthRight: "RIGHT_MOUTH",
        .leftShoulder: "LEFT_SHOULDER",
        .rightShoulder: "RIGHT_SHOULDER",
        .leftElbow: "LEFT_ELBOW",
        .rightElbow: "RIGHT_ELBOW",
        .leftWrist: "LEFT_WRIST",
        .rightWrist: "RIGHT_WRIST",
        .leftPinkyFinger: "LEFT_PINKY",
        .rightPinkyFinger: "RIGHT_PINKY",
        .leftIndexFinger: "LEFT_INDEX",
        .rightIndexFinger: "RIGHT_INDEX",
        .leftThumb: "LEFT_THUMB",
        .rightThumb: "RIGHT_THUMB",
        .leftHip: "LEFT_HIP",
        .rightHip: "RIGHT_HIP",
        .leftKnee: "LEFT_KNEE",
        .rightKnee: "RIGHT_KNEE",
        .leftAnkle: "LEFT_ANKLE",
        .rightAnkle: "RIGHT_ANKLE",
        .leftHeel: "LEFT_HEEL",
        .rightHeel: "RIGHT_HEEL",
        .leftToe: "LEFT_FOOT_INDEX",
        .rightToe: "RIGHT_FOOT_INDEX"
    ]

    let poses: [Pose]

    init(poses: [Pose]) {
        self.poses = poses
    }

    @objc public func toJSObject() -> AnyObject {
        var result = JSObject()
        result["poses"] = createPosesResult()
        return result as AnyObject
    }

    private func createPosesResult() -> JSArray {
        var result = JSArray()
        for pose in poses {
            result.append(createPoseResult(pose))
        }
        return result
    }

    private func createPoseResult(_ pose: Pose) -> JSObject {
        var result = JSObject()
        result["landmarks"] = createLandmarksResult(pose.landmarks)
        return result
    }

    private func createLandmarksResult(_ landmarks: [PoseLandmark]) -> JSArray {
        var result = JSArray()
        for landmark in landmarks {
            result.append(createLandmarkResult(landmark))
        }
        return result
    }

    private func createLandmarkResult(_ landmark: PoseLandmark) -> JSObject {
        var result = JSObject()
        result["type"] = ProcessImageResult.landmarkTypeMap[landmark.type] ?? ""
        result["x"] = Double(landmark.position.x)
        result["y"] = Double(landmark.position.y)
        result["z"] = Double(landmark.position.z)
        result["inFrameLikelihood"] = Double(landmark.inFrameLikelihood)
        return result
    }
}
